### PR TITLE
[Inductor] Improve reinplace_scatters pass

### PR DESCRIPTION
**Pull request by oulgen**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113090
* #113056
* __->__ #112801
* #113008
* #112752

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/_inductor/fx_passes/post_grad.py [hunk 1]
```diff
@@ -653,33 +656,36 @@ def reinplace_scatters(graph):
...
+                    continue
+                return True
+        return False
+
     def can_inplace(node, mutated_arg):
         if get_node_storage(mutated_arg) is None:
             return False
         shared_view_nodes = storage_to_nodes[get_node_storage(mutated_arg)]
-        if mutated_arg.op == "placeholder":
+        if any(view.op == "placeholder" for view in shared_view_nodes):
```
Comment (Chillee) [inline torch/_inductor/fx_passes/post_grad.py:676]: Why is this needed? It's ... plausibly correct to me, but did you find a case where it was needed?

Comment (oulgen) [reply torch/_inductor/fx_passes/post_grad.py:676]: Yes, there was a test case that was failing that used some index function which emitted an unsqueeze.

Comment (Chillee) [reply torch/_inductor/fx_passes/post_grad.py:676]: What was the exact graph being produced?

Comment (oulgen) [reply torch/_inductor/fx_passes/post_grad.py:676]: The gist is

arg0 <- graph input
buf1 = unsquueze arg0
buf2 = indexput buf1
buf3 = squueze buf2
return buf3

arg0 and buf1 share the memory location but we check whether buf1 is placeholder, so we inplace it, however since arg0 shares the same memory location it would be invalid to inplace unless there was a copy at the epilogue as well.

File: torch/_inductor/fx_passes/post_grad.py [hunk 2]
```diff
@@ -653,33 +656,36 @@ def reinplace_scatters(graph):
...
 
-            if len(shared_view_nodes) > 2:  # Arg aliases another node other than copy_
-                return False
-
-            # Check for any uses other than current node and copy_ epilogue
-            if len(mutated_arg.users) > 2:
+            if any_use_of_views_after_node(node, shared_view_nodes, copy_node):
                 return False
 
             graph.erase_node(copy_node)
```
[the remark this answers is not in the document]
Comment (oulgen) [reply torch/_inductor/fx_passes/post_grad.py:685]: We could move it but then we need to have a ternary return value which is slightly more confusing

File: torch/_inductor/fx_passes/post_grad.py (hunk 1 above)
Comment (Chillee) [reply torch/_inductor/fx_passes/post_grad.py:676]: Right, that makes sense. I think at some point when discussing this with you last week I thought of this case, but between then and now I forgot about it :P

**Review by Chillee — APPROVED** [review]
We should add a bit more correctness testing as well - the `test_perf` only tests the properties of the graph, but would be good to see some correctness.

**Review by jansel — APPROVED** [review]
